src/remote/action.go

```diff
@@ -0,0 +1,178 @@
+package remote
+
+import (
+	"encoding/hex"
+	"os"
+	"path"
+	"runtime"
+	"sort"
+	"strings"
+
+	pb "github.com/bazelbuild/remote-apis/build/bazel/remote/execution/v2"
+	"github.com/golang/protobuf/ptypes"
+
+	"github.com/thought-machine/please/src/core"
+)
+
+// uploadAction uploads a build action for a target and returns its digest.
+func (c *Client) uploadAction(target *core.BuildTarget, stamp []byte) (digest *pb.Digest, err error) {
+	err = c.uploadBlobs(func(ch chan<- *blob) error {
+		defer close(ch)
+		inputRoot, err := c.buildInputRoot(target, true)
+		if err != nil {
+			return err
+		}
+		inputRootDigest, inputRootMsg := digestMessageContents(inputRoot)
+		ch <- &blob{Data: inputRootMsg, Digest: inputRootDigest}
+		commandDigest, commandMsg := digestMessageContents(c.buildCommand(target, stamp))
+		ch <- &blob{Data: commandMsg, Digest: commandDigest}
+		action := &pb.Action{
+			CommandDigest:   commandDigest,
+			InputRootDigest: inputRootDigest,
+			Timeout:         ptypes.DurationProto(target.BuildTimeout),
+		}
+		actionDigest, actionMsg := digestMessageContents(action)
+		ch <- &blob{Data: actionMsg, Digest: actionDigest}
+		digest = actionDigest
+		return nil
+	})
+	return
+}
+
+// buildCommand builds the command for a single target.
+func (c *Client) buildCommand(target *core.BuildTarget, stamp []byte) *pb.Command {
+	return &pb.Command{
+		Platform: &pb.Platform{
+			Properties: []*pb.Platform_Property{
+				{
+					Name:  "OSFamily",
+					Value: translateOS(target.Subrepo),
+				},
+				// We don't really keep information around about ISA. Can look at adding
+				// that later if it becomes relevant & interesting.
+			},
+		},
+		// We have to run everything through bash since our commands are arbitrary.
```

Comment:
There's something in the docs about "toolchain" rules for things like bash, might be worth seeing how that stuff works?

Reply:
I'm not sure there's much we can do to communicate that - it mentions it around the `Platform` proto but all we can do there are communicate key-value pairs.

Will probably ask this one (and one or two others) on the mailing list at some point.